File: core/IntersectionOfTwoLinkedLists/IntersectionOfTwoLinkedLists.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def getIntersectionNode(headA: ListNode, headB: ListNode) -> ListNode:
    if headA is None or headB is None:
        return None
    remainA, remainB = headA, headB
    while headA != headB:
        headA = headB if headA is None else headA
        headB = headA if headB is None else headB

    return headA

File: core/IntersectionOfTwoLinkedLists/IntersectionOfTwoLinkedLists.java
package com.dbc;

public class IntersectionOfTwoLinkedLists {
    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) {
            return null;
        }
        ListNode remainA = headA, remainB = headB;
        while (headA != headB) {
            headA = headA == null ? remainB : headA.next;
            headB = headB == null ? remainA : headB.next;
        }
        return headA;
    }
}
